File: cpp/06_004.cpp
// Median-of-three pivot selection
template <typename T>
T median(const T& a, const T& b, const T& c) {
    if (a > b) {
        if (b > c) return b;      // a > b > c
        if (a > c) return c;      // a > c > b
        return a;                 // c > a > b
    } else {
        if (a > c) return a;      // b > a > c
        if (b > c) return c;      // b > c > a
        return b;                 // c > b > a
    }
}

template <typename T>
T choose_pivot(std::vector<T>& arr, int low, int high) {
    int mid = (low + high) / 2;
    return median(arr[low], arr[mid], arr[high]);
}

File: data structures and algorithms/cs/06_004.cs
/**
 * Selects a pivot using the median-of-three method.
 */
private int ChoosePivot(int[] arr, int low, int high)
{
   int mid = (low + high) / 2;

   // Find median of arr[low], arr[mid], and arr[high]
   int a = arr[low];
   int b = arr[mid];
   int c = arr[high];

   // Return the median value
   if (a > b)
   {
      if (b > c) return b;      // a > b > c
      else if (a > c) return c; // a > c > b
      else return a;            // c > a > b
   }
   else
   {
      if (a > c) return a;      // b > a > c
      else if (b > c) return c; // b > c > a
      else return b;            // c > b > a
   }
}
